src/widgets/helper/NotebookTab.hpp

```diff
@@ -107,6 +122,46 @@ class NotebookTab : public Button
 
     int normalTabWidthForHeight(int height) const;
 
+    bool shouldMessageHighlight(const ChannelView &channelViewSource,
+                                const MessagePtr &message) const;
+
+    struct HighlightSources {
+        struct ChannelViewProxy {
+            const ChannelView *channelView;
+        };
+
+        struct ChannelViewProxyHash {
+            using is_transparent = void;
+            std::size_t operator()(const ChannelViewProxy &cp) const noexcept;
+        };
+
+        struct ChannelViewProxyEqual {
+            bool operator()(const ChannelViewProxy &l,
+                            const ChannelViewProxy &r) const;
+        };
+
+        std::unordered_set<ChannelViewProxy, ChannelViewProxyHash,
+                           ChannelViewProxyEqual>
+            newMessageSource;
+        std::unordered_set<ChannelViewProxy, ChannelViewProxyHash,
+                           ChannelViewProxyEqual>
+            highlightedSource;
```

Comment:
Yes, I believe I see a possible issue. It might also crash on rehash, if those items no longer exist.

I believe I only require the
1) underlying channel
a) for comparison of channel == channel
b) for getName() -> identity (ID) of the source
2) channel view
a) for shouldIncludeMessage - uses the filter set (and the excludeUserMessagesFromFilter setting) - therefore it could work as a free function (provided we have the filter set)
b) getFilterIds -> identity (ID) of the source

Yes this should be replaced with some ID, I believe for the `ChannelView` rather than a `Channel`, granted it does not matter as long as the Name+FilterSet/FilterIds can be uniquely identified.

Reply:
Highlight source is just the split that is asking for a tab to be activated 
https://github.com/Chatterino/chatterino2/blob/e5e5a79645b7e3163468b0ae73ad75fbeb88ab00/src/widgets/splits/SplitContainer.cpp#L216-L224

Reply:
> Yes this should be replaced with some ID, I believe for the `ChannelView` rather than a `Channel`, granted it does not matter as long as the Name+FilterSet/FilterIds can be uniquely identified.

Yes I believe that for example a ChannelView could be generating these IDs, the sets would then just hold the IDs, as we actually do not do anything special with the ChannelView. I'm gonna try it out in the meantime to figure out where and how would be the best way.

Reply:
Replaced it with an ID based on those two things which now shouldnt suffer from any dangling pointers, (or the rehash problem due to them), as the ChannelView is only accessed when it is asked by the ChannelView (or the split) and IDs are saved at this access time, and the ChannelView is not stored anymore - https://github.com/Chatterino/chatterino2/pull/5649/commits/0e4897969b8e132f2f675c84f418c9412ac1d313

Reply:
As the ID was being asked for with every highlight request (i.e. all messages that ultimately appear in a currently unviewed tab's splits), I moved the ID into the `ChannelView`, it is updated when filters are updated or the underlying channel is updated. https://github.com/Chatterino/chatterino2/pull/5649/commits/1bdb11839129fe3ceff801746b653dc77f4d4622.